File: app/src/main/java/beg/hr/rxredux/java/redux/presentation/ViewDriver.java
package beg.hr.rxredux.java.redux.presentation;

import android.support.v4.view.ViewCompat;
import android.view.View;

import com.jakewharton.rxbinding.view.RxView;
import com.jakewharton.rxbinding.view.ViewAttachEvent;
import com.jakewharton.rxbinding.view.ViewAttachEvent.Kind;
import com.trello.rxlifecycle.LifecycleProvider;
import com.trello.rxlifecycle.LifecycleTransformer;
import com.trello.rxlifecycle.RxLifecycle;
import com.trello.rxlifecycle.android.ActivityEvent;
import com.trello.rxlifecycle.android.RxLifecycleAndroid;

import javax.annotation.CheckReturnValue;
import javax.annotation.Nonnull;

import beg.hr.rxredux.java.redux.Command;
import rx.Observable;

import static rx.Observable.merge;

/** Created by juraj on 17/03/2017. */
public abstract class ViewDriver<V extends View, S> implements LifecycleProvider<ViewAttachEvent> {
  private final V view;
  private final ViewAttachEvent attachEvent;
  private final ViewAttachEvent detachEvent;
  private final LifecycleProvider<ActivityEvent> lifecycleProvider;

  protected ViewDriver(V view, LifecycleProvider<ActivityEvent> lifecycleProvider) {
    this.view = view;
    this.lifecycleProvider = lifecycleProvider;
    this.attachEvent = ViewAttachEvent.create(view, Kind.ATTACH);
    this.detachEvent = ViewAttachEvent.create(view, Kind.DETACH);
  }

  public abstract void input(S state);

  public abstract Observable<Command> output();

  public V view() {
    return view;
  }

  private boolean isAttach(ViewAttachEvent e) {
    return e.equals(attachEvent);
  }

  private boolean isDetach(ViewAttachEvent e) {
    return e.equals(detachEvent);
  }

  @Nonnull
  @Override
  public Observable<ViewAttachEvent> lifecycle() {
    Observable<ViewAttachEvent> events = RxView.attachEvents(this.view);
    if (ViewCompat.isAttachedToWindow(view)) events = events.startWith(attachEvent);
    return events;
  }

  @Nonnull
  @Override
  public <T> LifecycleTransformer<T> bindUntilEvent(ViewAttachEvent event) {
    return RxLifecycle.bindUntilEvent(lifecycle(), event);
  }

  @Nonnull
  @Override
  public <T> LifecycleTransformer<T> bindToLifecycle() {
    return RxLifecycleAndroid.bindView(this.view);
  }

  @Nonnull
  @CheckReturnValue
  public <T> LifecycleTransformer<T> bindUntilDetach() {
    return bindUntilEvent(ViewAttachEvent.create(view, Kind.DETACH));
  }

  public Observable<Boolean> load() {
    Observable<Boolean> attachEvent =
        lifecycle().filter(this::isAttach).map(viewAttachEvent -> true);
    Observable<Boolean> onStartEvent =
        lifecycleProvider
            .lifecycle()
            .filter(activityEvent -> activityEvent.equals(ActivityEvent.START))
            .map(activityEvent -> true);

    Observable<Boolean> detachEvent =
        lifecycle().filter(this::isDetach).map(viewAttachEvent -> false);
    Observable<Boolean> onStopEvent =
        lifecycleProvider
            .lifecycle()
            .filter(activityEvent -> activityEvent.equals(ActivityEvent.STOP))
            .map(activityEvent -> false);

    return merge(attachEvent, onStartEvent, detachEvent, onStopEvent)
        .distinctUntilChanged()
        .compose(bindUntilDetach())
        .filter(Boolean::booleanValue);
  }

  public <T> LifecycleTransformer<T> bindUntilDestroy() {
    // TODO: 17/03/2017 add onStop
    return bindUntilDetach();
  }
}

File: app/src/main/java/beg/hr/rxredux/kotlin/util/ui/ViewDriver.kt
package beg.hr.rxredux.kotlin.util.ui

import android.support.v4.view.ViewCompat
import android.view.View
import com.jakewharton.rxbinding.view.RxView
import com.jakewharton.rxbinding.view.ViewAttachEvent
import com.trello.rxlifecycle.LifecycleProvider
import com.trello.rxlifecycle.LifecycleTransformer
import com.trello.rxlifecycle.RxLifecycle
import com.trello.rxlifecycle.android.ActivityEvent
import com.trello.rxlifecycle.android.RxLifecycleAndroid
import rx.Observable
import rx.Observable.merge
import rx.subjects.BehaviorSubject

/**
 * Created by juraj on 24/03/2017.
 */
abstract class ViewDriver<out V : View, in S, C>(val view: V, val lifecycleProvider: LifecycleProvider<ActivityEvent>) : LifecycleProvider<ViewAttachEvent> {

    val attachEvent: ViewAttachEvent = ViewAttachEvent.create(view, ViewAttachEvent.Kind.ATTACH)
    val detachEvent: ViewAttachEvent = ViewAttachEvent.create(view, ViewAttachEvent.Kind.DETACH)

    fun loadEvent(): Observable<Boolean> {
        val activityEvents = BehaviorSubject.create<ActivityEvent>()

        lifecycleProvider.lifecycle().subscribe({ activityEvents.onNext(it) })
        val lifecycle = activityEvents.asObservable().share()

        val attachEvent = lifecycle()
                .filter({ this.isAttach(it) })
                .map { true }
                .filter {
                    // fixme sometimes attach event is called after onStop ?!
                    // Don't let it through if last activity event is stop
                    ActivityEvent.STOP != activityEvents.value
                }

        val onStartEvent = lifecycle
                .filter { activityEvent -> activityEvent == ActivityEvent.START }
                .map { true }

        val detachEvent = lifecycle().filter({ this.isDetach(it) }).map { false }

        val onStopEvent = lifecycle
                .filter { activityEvent -> activityEvent == ActivityEvent.STOP }
                .map { false }

        return merge(attachEvent, onStartEvent, detachEvent, onStopEvent)
                .distinctUntilChanged()
                .compose(bindUntilDetach())
                .filter { it }
    }

    fun isAttach(e: ViewAttachEvent): Boolean {
        return e == attachEvent
    }

    fun isDetach(e: ViewAttachEvent): Boolean {
        return e == detachEvent
    }

    override fun lifecycle(): Observable<ViewAttachEvent> {
        var events: Observable<ViewAttachEvent> = RxView.attachEvents(view)
        if (ViewCompat.isAttachedToWindow(view)) events = events.startWith(attachEvent)
        return events
    }

    override fun <T : Any?> bindUntilEvent(event: ViewAttachEvent): LifecycleTransformer<T> = RxLifecycle.bindUntilEvent(lifecycle(), event)

    override fun <T : Any?> bindToLifecycle(): LifecycleTransformer<T> = RxLifecycleAndroid.bindView(view)

    fun <T> bindUntilDetach(): LifecycleTransformer<T> = bindUntilEvent(ViewAttachEvent.create(view, ViewAttachEvent.Kind.DETACH))

    abstract fun render(state: S)

    abstract fun commands(): Observable<C>
}
